### PR TITLE
✨ Add v1alpha6 API types

### LINKED ISSUE #1270: Add v1alpha6 API types
/kind feature

Now that v0.6 with the new API version v1alpha5 is released, we want to create a new API Version `v1alpha6`. This API Version will be published with the next CAPO version v0.7.

### PR DESCRIPTION
**Pull request by tobiasgiese**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1270

- [x] copy previous API (`v1alpha5`) to new API (`v1alpha6`)
- [x] remove annotation `+kubebuilder:storageversion` from previous API
- [x] remove webhooks from previous API
- [x] add conversion to previous APIs `doc.go` to enable conversion generation
- [x] impl conversion to previous API -- take care of the marhsalling (see #1247 for more info)
- [x] add new API dir (`--input-dirs`) to the `generate-go` makefile target
- [ ] ~~add new contract to `metadata.yaml` (top dir and e2e tests)~~ (only on new relase)
- [x] add new API contract to `config/crd/kustomization.yaml` (see [Apply the contract version label](https://cluster-api.sigs.k8s.io/developer/providers/v1alpha2-to-v1alpha3.html#apply-the-contract-version-label-clusterx-k8sioversion-version1_version2_version3-to-your-crds) docs in the CAPI book, must be `_` separated, if we have multiple API versions for the same CAPI version)
- [x] make use of new API in our controllers and tests (i.e., all files)
- [x] add scheme of new API to `main.go`
- [x] update version in `test/e2e/data/e2e_conf.yaml`
- [x] update readme

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**TODOs**:

- [x] squashed commits
- if necessary:
  - [ ] includes documentation
  - [ ] adds unit tests

/hold

<sub>Tobias Giese <tobias.giese@mercedes-benz.com>, Mercedes-Benz Tech Innovation GmbH, [legal info/Impressum](https://github.com/mercedes-benz/foss/blob/master/LEGAL_IMPRINT.md)</sub>

### CODE REVIEW

File: api/v1alpha6/conversion.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
+/*
+Copyright 2021 The Kubernetes Authors.
```
Comment (seanschneeweiss) [inline api/v1alpha6/conversion.go:2]: I don't know the policy, but should we use the year of file creation?
```suggestion
Copyright 2022 The Kubernetes Authors.
```

**Review by seanschneeweiss — COMMENTED** [review]
Thanks Tobias for creating the v1alpha6 :+1:

File: api/v1alpha6/openstackcluster_types.go [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+	// Any transient errors that occur during the reconciliation of
+	// OpenStackClusters can be added as events to the OpenStackCluster object
+	// and/or logged in the controller's output.
+	// +optional
+	FailureMessage *string `json:"failureMessage,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=openstackclusters,scope=Namespaced,categories=cluster-api,shortName=osc
+// +kubebuilder:storageversion
```
Comment (seanschneeweiss) [inline api/v1alpha6/openstackcluster_types.go:210]: As v1alpha6 should be the default for storing the data, `// +kubebuilder:storageversion` has to be removed from the v1alpha5 files.

Comment (tobiasgiese) [reply api/v1alpha6/openstackcluster_types.go:210]: Ah, totally must missed that yesterday. Thanks for the hint :)

File: api/v1alpha5/conversion.go [hunk 3]
```diff
@@ -16,26 +16,106 @@ limitations under the License.
...
+var _ ctrlconversion.Convertible = &OpenStackCluster{}
 
-// Hub marks OpenStackClusterTemplateList as a conversion hub.
-func (*OpenStackClusterTemplateList) Hub() {}
+func (r *OpenStackCluster) ConvertTo(dstRaw ctrlconversion.Hub) error {
+	dst := dstRaw.(*infrav1.OpenStackCluster)
 
-// Hub marks OpenStackMachine as a conversion hub.
-func (*OpenStackMachine) Hub() {}
+	return Convert_v1alpha5_OpenStackCluster_To_v1alpha6_OpenStackCluster(r, dst, nil)
```
Comment (bavarianbidi) [inline api/v1alpha5/conversion.go:30]: For not loosing information during conversion, we have to unmarshal the data to get it added to the `cluster.x-k8s.io/conversion-data` Annotation (see current implementation in `v1alpha4` - https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/e78f0b83b0297590043826138b33f7c050f1ba53/api/v1alpha4/conversion.go#L32-L44)

This must be also done in the `ConvertFrom` method below (see current implementation in `v1alpha4` - https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/e78f0b83b0297590043826138b33f7c050f1ba53/api/v1alpha4/conversion.go#L52-L59)

This procedure applies also to `ConvertTo`/`ConvertFrom` methods for `OpenStackMachine` and `OpenStackMachineTemplate`.

Unfourtanetly i just discovered, that i didn't implemented the `utilconversion.UnmarshalData` in `ConvertTo` for `OpenStackMachine` and `OpenStackMachineTemplate` when i introduced this in https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/1247.

File: api/v1alpha5/conversion.go [hunk 4]
```diff
@@ -16,26 +16,106 @@ limitations under the License.
...
+}
 
-// Hub marks OpenStackMachineList as a conversion hub.
-func (*OpenStackMachineList) Hub() {}
+func (r *OpenStackCluster) ConvertFrom(srcRaw ctrlconversion.Hub) error {
+	src := srcRaw.(*infrav1.OpenStackCluster)
 
-// Hub marks OpenStackMachineTemplate as a conversion hub.
-func (*OpenStackMachineTemplate) Hub() {}
+	return Convert_v1alpha6_OpenStackCluster_To_v1alpha5_OpenStackCluster(src, r, nil)
```
Comment (bavarianbidi) [inline api/v1alpha5/conversion.go:36]: see comment above

File: api/v1alpha5/conversion.go [hunk 5]
```diff
@@ -16,26 +16,106 @@ limitations under the License.
...
+
+	return Convert_v1alpha6_OpenStackClusterTemplate_To_v1alpha5_OpenStackClusterTemplate(src, r, nil)
+}
+
+var _ ctrlconversion.Convertible = &OpenStackMachine{}
+
+func (r *OpenStackMachine) ConvertTo(dstRaw ctrlconversion.Hub) error {
+	dst := dstRaw.(*infrav1.OpenStackMachine)
+
+	return Convert_v1alpha5_OpenStackMachine_To_v1alpha6_OpenStackMachine(r, dst, nil)
```
Comment (bavarianbidi) [inline api/v1alpha5/conversion.go:72]: see comment above

File: api/v1alpha5/conversion.go [hunk 6]
```diff
@@ -16,26 +16,106 @@ limitations under the License.
...
+}
+
+func (r *OpenStackMachine) ConvertFrom(srcRaw ctrlconversion.Hub) error {
+	src := srcRaw.(*infrav1.OpenStackMachine)
+
+	return Convert_v1alpha6_OpenStackMachine_To_v1alpha5_OpenStackMachine(src, r, nil)
```
Comment (bavarianbidi) [inline api/v1alpha5/conversion.go:78]: see comment above

File: api/v1alpha5/conversion.go [hunk 7]
```diff
@@ -16,26 +16,106 @@ limitations under the License.
...
+func (r *OpenStackMachineTemplate) ConvertTo(dstRaw ctrlconversion.Hub) error {
+	dst := dstRaw.(*infrav1.OpenStackMachineTemplate)
+
+	return Convert_v1alpha5_OpenStackMachineTemplate_To_v1alpha6_OpenStackMachineTemplate(r, dst, nil)
+}
+
+func (r *OpenStackMachineTemplate) ConvertFrom(srcRaw ctrlconversion.Hub) error {
+	src := srcRaw.(*infrav1.OpenStackMachineTemplate)
+
+	return Convert_v1alpha6_OpenStackMachineTemplate_To_v1alpha5_OpenStackMachineTemplate(src, r, nil)
```
Comment (bavarianbidi) [inline api/v1alpha5/conversion.go:106]: see comment above

File: api/v1alpha5/conversion.go [hunk 8]
```diff
@@ -16,26 +16,106 @@ limitations under the License.
...
+
+	return Convert_v1alpha6_OpenStackMachineList_To_v1alpha5_OpenStackMachineList(src, r, nil)
+}
+
+var _ ctrlconversion.Convertible = &OpenStackMachineTemplate{}
+
+func (r *OpenStackMachineTemplate) ConvertTo(dstRaw ctrlconversion.Hub) error {
+	dst := dstRaw.(*infrav1.OpenStackMachineTemplate)
+
+	return Convert_v1alpha5_OpenStackMachineTemplate_To_v1alpha6_OpenStackMachineTemplate(r, dst, nil)
```
Comment (bavarianbidi) [inline api/v1alpha5/conversion.go:100]: see comment above

**Review by bavarianbidi — COMMENTED** [review]
For having a working `hub-spoke-hup` / `spoke

File: api/v1alpha5/conversion.go (hunk 3 above)
Comment (tobiasgiese) [reply api/v1alpha5/conversion.go:30]: That's an interesting and nasty thing, tbh. Are we able to unit test this somehow? What data is going to be lost without the (un)marshalling?

Comment (tobiasgiese) [reply api/v1alpha5/conversion.go:30]: That is really a nasty thing, you're toally right. Thanks! Let me implement this for these 3 types.

Comment (tobiasgiese) [reply api/v1alpha5/conversion.go:30]: Are we able to test this somehow?

Comment (bavarianbidi) [reply api/v1alpha5/conversion.go]: > Are we able to test this somehow?

if you already have a cluster where `v1alpha6` is applied, you could do a `k get openstackcluster.v1alpha5.infrastructure.cluster.x-k8s.io -o yaml`. This should give you beside the spec of the object in version `v1alpha5` in the `cluster.x-k8s.io/conversion-data` annotation the same spec in version `v1alpha6`.

Comment (tobiasgiese) [reply api/v1alpha5/conversion.go]: I also added the marshalling fix for `OpenStackClusterTemplate`

File: config/crd/kustomization.yaml [hunk 9]
```diff
@@ -1,7 +1,7 @@
 commonLabels:
   cluster.x-k8s.io/v1alpha3: v1alpha3
   cluster.x-k8s.io/v1alpha4: v1alpha4
-  cluster.x-k8s.io/v1beta1: v1alpha5
+  cluster.x-k8s.io/v1beta1: v1alpha5_v1alpha6
```
Comment (apricote) [inline config/crd/kustomization.yaml:4]: Our v1alpha4 API is also compatible with CAPI v1beta1, and was used as such in the v0.5 release. Should we also include this here?

File: metadata.yaml [hunk 10]
```diff
@@ -13,3 +13,6 @@ releaseSeries:
 - major: 0
   minor: 6
   contract: v1beta1
+- major: 0
```
Comment (apricote) [inline metadata.yaml:16]: This change is technically not related to a new API Version, but to a new release series. Should we add this now or right before cutting a release branch?

File: config/crd/kustomization.yaml (hunk 9 above)
Comment (tobiasgiese) [reply config/crd/kustomization.yaml:4]: I'd say we must set this, yes. See the [Apply the contract version label](https://cluster-api.sigs.k8s.io/developer/providers/v1alpha2-to-v1alpha3.html#apply-the-contract-version-label-clusterx-k8sioversion-version1_version2_version3-to-your-crds) docs in the CAPI book

File: metadata.yaml (hunk 10 above)
Comment (tobiasgiese) [reply metadata.yaml:16]: I'm fine with both and have also not a hard opinion on this.
If we add it now, the `clusterctl` request for a CAPO 0.7 release will fail anyway as we haven't released it yet.

File: config/crd/kustomization.yaml (hunk 9 above)
Comment (tobiasgiese) [reply config/crd/kustomization.yaml:4]: We also have to update our [Readme](https://github.com/kubernetes-sigs/cluster-api-provider-openstack/#compatibility-with-cluster-api-and-kubernetes-versions) in this case

File: metadata.yaml (hunk 10 above)
Comment (tobiasgiese) [reply metadata.yaml:16]: But as we maybe want to graduate this version to v1beta1 it‘s better to remove it

**Comment by apricote** [comment]
I checked out the branch locally (1e96a47f823f9f901e981342ca1843544ad06e52) and discovered an error when running `make generate-manifests`:

```
➜ cluster-api-provider-openstack git:(v1alpha6) git describe          
v0.6.0-39-g1e96a47f
➜ cluster-api-provider-openstack git:(v1alpha6) make generate-manifests
hack/tools/bin/controller-gen \
        paths=./api/... \
        crd:crdVersions=v1 \
        output:crd:dir=config/crd/bases \
        output:webhook:dir=config/webhook \
        webhook
hack/tools/bin/controller-gen \
        paths=./controllers/... \
        output:rbac:dir=config/rbac \
        rbac:roleName=manager-role
panic: interface conversion: types.Type is nil, not *types.Named

goroutine 1 [running]:
sigs.k8s.io/controller-tools/pkg/crd.namedToSchema(0xc0032b6ae0, 0xc000c79188)
        /home/jt/go/pkg/mod/sigs.k8s.io/controller-tools@v0.7.0/pkg/crd/schema.go:248 +0x318
sigs.k8s.io/controller-tools/pkg/crd.typeToSchema(0xc0032b6ae0, {0xd3f790?, 0xc000c79188})
        /home/jt/go/pkg/mod/sigs.k8s.io/controller-tools@v0.7.0/pkg/crd/schema.go:170 +0xcd
sigs.k8s.io/controller-tools/pkg/crd.structToSchema(0xc0049ea3e8, 0xc000c791b8)
        /home/jt/go/pkg/mod/sigs.k8s.io/controller-tools@v0.7.0/pkg/crd/schema.go:393 +0x77a
sigs.k8s.io/controller-tools/pkg/crd.typeToSchema(0xc0049ea3e8, {0xd3f850?, 0xc000c791b8})
        /home/jt/go/pkg/mod/sigs.k8s.io/controller-tools@v0.7.0/pkg/crd/schema.go:178 +0x12f
sigs.k8s.io/controller-tools/pkg/crd.infoToSchema(0xc0049ea3e8)
        /home/jt/go/pkg/mod/sigs.k8s.io/controller-tools@v0.7.0/pkg/crd/schema.go:121 +0x10e
sigs.k8s.io/controller-tools/pkg/crd.(*Parser).NeedSchemaFor(0xc000033ec0, {0xc00030da60?, {0xc002342720?, 0x7?}})
        /home/jt/go/pkg/mod/sigs.k8s.io/controller-tools@v0.7.0/pkg/crd/parser.go:190 +0x2b7
sigs.k8s.io/controller-tools/pkg/crd.(*Parser).NeedFlattenedSchemaFor(0xc000033ec0, {0xc00030da60?, {0xc002342720?, 0xc164fe?}})
        /home/jt/go/pkg/mod/sigs.k8s.io/controller-tools@v0.7.0/pkg/crd/parser.go:202 +0xb7
sigs.k8s.io/controller-tools/pkg/crd.(*Parser).NeedCRDFor(0xc000033ec0, {{0xc0004672fe?, 0x4de8496d10?}, {0xc002342720?, 0xd0?}}, 0x0)
        /home/jt/go/pkg/mod/sigs.k8s.io/controller-tools@v0.7.0/pkg/crd/spec.go:85 +0x56e
sigs.k8s.io/controller-tools/pkg/crd.Generator.Generate({0x0, 0x0, {0xc00020ae70, 0x1, 0x1}, 0x0}, 0xc0000b00f0)
        /home/jt/go/pkg/mod/sigs.k8s.io/controller-tools@v0.7.0/pkg/crd/gen.go:118 +0x2da
sigs.k8s.io/controller-tools/pkg/genall.(*Runtime).Run(0xc0002a5560)
        /home/jt/go/pkg/mod/sigs.k8s.io/controller-tools@v0.7.0/pkg/genall/genall.go:213 +0x246
main.main.func1(0xc00024e500?, {0xc0000b1ea0?, 0x5?, 0x5?})
        /home/jt/go/pkg/mod/sigs.k8s.io/controller-tools@v0.7.0/cmd/controller-gen/main.go:176 +0x6a
github.com/spf13/cobra.(*Command).execute(0xc00024e500, {0xc0000321f0, 0x5, 0x5})
        /home/jt/go/pkg/mod/github.com/spf13/cobra@v1.4.0/command.go:856 +0x67c
github.com/spf13/cobra.(*Command).ExecuteC(0xc00024e500)
        /home/jt/go/pkg/mod/github.com/spf13/cobra@v1.4.0/command.go:974 +0x3b4
github.com/spf13/cobra.(*Command).Execute(...)
        /home/jt/go/pkg/mod/github.com/spf13/cobra@v1.4.0/command.go:902
main.main()
        /home/jt/go/pkg/mod/sigs.k8s.io/controller-tools@v0.7.0/cmd/controller-gen/main.go:200 +0x2c6
make: *** [Makefile:240: generate-manifests] Error 2
```

Can you replicate this @tobiasgiese?

**Comment by tobiasgiese** [comment]
> /lgtm

Had to downgrade the dev test version in our e2e_conf.yaml as I removed the 0.7 contract

/hold cancel

**Pull request merged**